V0.82 - Add support for binary mmCIF (BCIF) reading and writing in IoAdapter

piehld opened the pull request:
Add support for binary mmCIF (BCIF) reading and writing in IoAdapterPy and IoAdapterCore

piehld commented:
There appears to be an issue with the `IoAdapterCore` module and reading in dictionary files. Specifically, the `mmcif_pdbx_v5_next.dic` dictionary is reporting 6798 errors (all but 2 of which are syntax errors):

```
2023-10-30 19:43:16,916 [ERROR]-IoAdapterCore.__processReadLogFile: /py-mmcif/mmcif/tests/test-output/mmcif_pdbx_v5_next_cif-parser-ascii_P16987093946821_701.cif syntax errors 6796  all errors 6798
```

When I print out the two non-syntax errors, this is what the message says:
```
testDictReaderUtf8 (__main__.IoAdapterTests) ... 
2023-10-30 19:50:46,545 [INFO]-IoAdapterCore.__processReadLogFile: 'ERROR - Number of data values is not exact multiples of the number of data names   (look above line 6508)'
2023-10-30 19:50:46,548 [INFO]-IoAdapterCore.__processReadLogFile: 'ERROR - Number of data values is not exact multiples of the number of data names   (look above line 19715)'
2023-10-30 19:50:46,593 [ERROR]-IoAdapterCore.__processReadLogFile: /py-mmcif/mmcif/tests/data/mmcif_pdbx_v5_next.dic syntax errors 6796  all errors 6798
```

Apparently this has been a longstanding issue, as previously the dictionary tests were commented out. However, if we want to use the dictionary-based typing for BCIF encoding, we will need the `IoAdapaterCore` to be able to read in a dictionary file.

Interestingly, this is only an issue for the `IoAdapaterCore` module. The `IoAdapaterPy` module works is able to read in the dictionary just fine.

epeisach commented:
I have reviewed your changes.  The only thing you might want to consider is what if fmt specified is not mmcif or bcif?  You will be returned an empty list all the time.  No exceptions or logged errors.  Is this the desired situation?

(still looking at dictionary)

epeisach commented:
I see that you re-enabled the dictionary parsing with IoAdapter (i.e. IoAdapterCore).   There is a reason why John's code always uses the IoAdapterPy.

That said, IoAdapterCore should be able to support --- but the C++ bindings are probably not supporting. 

cpp-cif-parser/src/DICParserBase.C DicParser is to be used instead of CifParser.

Python C++ IoAdapterCore bindings do not support dictionary.  If we would like to add C++ support for the Dictionary parser - it already exists in the C++ bindings in mmciflib but we need a different wrapper. 

    from mmcif.core.mmciflib import ParseCifSimple, CifFile, ParseCifSelective, CifFileReadDef
Are all Cif based.  -- not Star based.

So - I suggest disable those tests.  If we would like to add a C++ DicParser - we should add that separately.

epeisach commented:
Updating to the latest submodules should be safe.  But will not solve this issue. Dictionary parsing from C++ is a different API.  It so happens that the IoAdapterPy manages ok.

piehld commented:
Thanks for the review and insight, @epeisach!

> So - a quick read through your changes -- you are essentially extending the IoAdapterCore and IoAdapterPy to support extra parameters to specify if the binary reader/writer should be used. No change to the writer is made. So essentially moving data over for a conditional.

Yes basically, which will make it easier to be used by our [rcsb.utils.io.MarshalUtil](https://github.com/rcsb/py-rcsb_utils_io/blob/master/rcsb/utils/io/MarshalUtil.py).

> The only thing you might want to consider is what if fmt specified is not mmcif or bcif? You will be returned an empty list all the time. No exceptions or logged errors. Is this the desired situation?

Good point, I'll add a warning about that. Thanks.

> I see that you re-enabled the dictionary parsing with IoAdapter (i.e. IoAdapterCore). There is a reason why John's code always uses the IoAdapterPy.

Actually, most of John's code appears to use the `IoAdapterCore` by default (e.g., see the import section in [rcsb.utils.io.IoUtil](https://github.com/rcsb/py-rcsb_utils_io/blob/master/rcsb/utils/io/IoUtil.py#L67-L70)). This is why I was attempting here to add the BCIF support to `IoAdapterCore` as well. Otherwise, if his code was only using the `IoAdapterPy`, then yeah everything works fine.

I'm not sure why he opted to default to `IoAdapterCore` in that code, but I'm guessing it had to do with performance. I suppose at this point it's worth doing a full evaluation of the impact on performance with using `IoAdapterPy` instead. If it's minimal, then we could [at least for now] just add the BCIF  support to `IoAdapterPy` and try to tackle `IoAdapterCore` later, since it sounds like creating a new wrapper with C++ dictionary parser support will be pretty involved.

epeisach commented:
Mind if I add the one line fix to produce MacOS arm64 packages? (as well as newer python)

epeisach commented:
I put it in here.  You have already bumped the version in the pull request - so should go there.

piehld commented:
@epeisach Thanks for all the updates and fixes. 

As I described offline, I was able to run the performance comparison between using `IoAdapterCore` vs `IoAdapterPy` in loading our exchange DB, and determined that there wasn't any significant change in the loading speed. As such, there's no longer an major demand to provide BCIF support for `IoAdapterCore`, since we can use `IoAdapterPy` instead.  Of course, if there arises a large demand for that capability we can return to it and assign it a new priority, but given the amount of work that will be required (i.e., creating the Python wrapper for the C++ dictionary parser) I think we can lower the priority for now.

With this in mind, I reverted most of the changes to `IoAdapterCore` while keeping the additions in `IoAdapterPy`, along with a variety of other minor improvements. I also reverted the disabling of the testing that was required earlier due to issues with `urllib3`, since it no longer crops up after the revisions to `IoAdapterCore`.

Last, I also merged the other long-standing PR, which provides a very simple and valuable fix.
